crates/hir-ty/src/next_solver/infer/canonical/canonicalizer.rs

```diff
@@ -384,10 +390,9 @@ impl<'cx, 'db> TypeFolder<DbInterner<'db>> for Canonicalizer<'cx, 'db> {
                             // FIXME: perf problem described in #55921.
                             ui = UniverseIndex::ROOT;
                         }
-                        self.canonicalize_ty_var(
-                            CanonicalVarKind::Ty(CanonicalTyVarKind::General(ui)),
-                            t,
-                        )
+
+                        let sub_root = self.get_or_insert_sub_root(vid);
+                        self.canonicalize_ty_var(CanonicalVarKind::Ty { ui, sub_root }, t)
```

Comment:
existing: why does this canonicalizer exists. The next solver canonicalizer is in `rustc_next_trait_solver`. I guess you need this canonicalizer outside of the trait solver :thinking:  where?